python/pypto/language/distributed/op/tensor_ops.py

```
@@ -559,11 +561,18 @@ def allreduce(
-    a fresh row, so active non-self cells end at non-zero values while the
-    skipped self column remains zero. The monotonic advance is per-row, not
-    per-cell reuse.
+    Ring mode first views a packed ND target as one linear stream, then
+    traverses each segment in physical subchunks of at most 16 KiB. FP32 uses
+    balanced logical segment boundaries. FP16 aligns every non-empty segment
+    start and remote tail span to 32 bytes while restoring the logical
+    ``valid_shape`` before reduction and store. This also covers ``SIZE < NR``
+    without changing the packed public layout or dropping elements.
+
+    **Ring barrier protocol:** each ring round owns one row of the
+    ``[2*(NR−1), NR]`` signal. Every subchunk advances that row twice:
```

Ruff (0.15.21)

[warning] Docstring contains ambiguous `−` (MINUS SIGN). Did you mean `-` (HYPHEN-MINUS)?

(RUF002)